### PR TITLE
GemmConvMobileFunction(optimized for mobile)

### LINKED ISSUE #6801: Memory optimization in convolution layer.
For the large input images, the convolution layer by the implementation of im2col and gemm will allocate a large size of workspace for the process of im2col. For example, in one OCR based detection model, the input image size is 3 * 1500 * 1500 and the first convolution layer is:

```python
    conv1 = paddle.layer.img_conv(
            input=data,   # 3 * 1500 * 1500
            filter_size=7,
            num_channels=3,
            num_filters=32,
            stride=4,
            padding=3,
            act=paddle.activation.Relu(),
            bias_attr=None)
```

The output shape of this convolution layer is 32 * 375 * 375. The size of  extra workspace for im2col of this layer is 78 MB:

```cpp
// Ci * Kh * Kw * output_height * output_width.
3 * 7 * 7 * 375 * 375 * sizeof(float) / 1024 / 1024 = 78MB
```

If more convolution layer like this in the network will lead to more memory usage. Especially for the mobile deployment, this is a very large memory . So we need to optimize it. We adopt the groped `im2col + gemm` to calculate the convolution layer with large extra workspace.

### PR DESCRIPTION
**Pull request by hedaoyuan**
Fix #6801 #6802 
https://github.com/PaddlePaddle/Paddle/pull/6802#pullrequestreview-85463693

### CODE REVIEW

File: paddle/function/GemmConvOp.cpp [hunk 1]
```diff
@@ -126,14 +126,163 @@ class GemmConvFunction : public ConvFunctionBase {
...
+        TensorShape({inputChannels / groups_, inputHeight, inputWidth});
+
+    TensorShape colShape;
+    real* colData = NULL;
+
+    size_t colHeight = inputChannels / groups_ * filterHeight * filterWidth;
+    size_t colWidth = outputHeight * outputWidth;
+    // Max col matrix height 256, Max col matrix width 1024
+    size_t stepColHeight = std::min(colHeight, (size_t)256);
+    size_t stepColWidth = std::min(colWidth, (size_t)2048);
```
Comment (qingqing01) [inline paddle/function/GemmConvOp.cpp]:
```cpp
static_cast<size_t>(256)
static_cast<size_t>(2048)
```

File: paddle/function/Im2ColTest.cpp [hunk 2]
```diff
@@ -138,4 +138,86 @@ TEST(Im2ColFunctor, GPU) { TestIm2ColFunctor<DEVICE_TYPE_GPU, float>(); }
...
+template <class T>
+void TestIm2ColMobileFunctor() {
+  for (size_t channels : {1, 5, 32}) {
+    for (size_t inputHeight : {5, 33, 100}) {
+      for (size_t inputWidth : {5, 32, 96}) {
+        for (size_t filterHeight : {1, 5}) {
+          for (size_t filterWidth : {3, 7}) {
+            for (size_t stride : {1, 2}) {
+              for (size_t padding : {0, 1}) {
+                for (size_t dilation : {1, 3}) {
```
Comment (qingqing01) [inline paddle/function/Im2ColTest.cpp]: Maybe the test case can be reduced to speed up the unit testing time.

File: paddle/function/Im2Col.h [hunk 3]
```diff
@@ -98,4 +98,54 @@ class Col2ImFunctor {
...
+        int h = (colWidthStart + colw) / outputWidth;
+        int w = (colWidthStart + colw) % outputWidth;
+
+        int imRowIdx = h * strideHeight + hOffset * dilationHeight;
+        int imColIdx = w * strideWidth + wOffset * dilationWidth;
+        if ((imRowIdx - paddingHeight) < 0 ||
+            (imRowIdx - paddingHeight) >= inputHeight ||
+            (imColIdx - paddingWidth) < 0 ||
+            (imColIdx - paddingWidth) >= inputWidth) {
+          colData[colh * colWidthSize + colw] = T(0);
```
Comment (qingqing01) [inline paddle/function/Im2Col.h]: ```static_cast<T>(0)```

File: paddle/function/GemmConvOp.cpp [hunk 4]
```diff
@@ -126,14 +126,163 @@ class GemmConvFunction : public ConvFunctionBase {
...
+
+    if (needIm2col) {
+      colShape = TensorShape({inputChannels / groups_,
+                              filterHeight,
+                              filterWidth,
+                              outputHeight,
+                              outputWidth});
+
+      resizeBuffer<Device>(stepColHeight * stepColWidth * sizeof(real));
+      colData = reinterpret_cast<real*>(memory_->getBuf());
```
Comment (qingqing01) [inline paddle/function/GemmConvOp.cpp]: I see that the `memory_` is not released in every layer. Do you think the if at most 2M extra workspace is used in every layer, we do not need to release it?

File: paddle/function/GemmConvOp.cpp (hunk 1 above)
Comment (hedaoyuan) [reply paddle/function/GemmConvOp.cpp]: Done.

File: paddle/function/Im2Col.h (hunk 3 above)
Comment (hedaoyuan) [reply paddle/function/Im2Col.h]: Done.

File: paddle/function/Im2ColTest.cpp (hunk 2 above)
Comment (hedaoyuan) [reply paddle/function/Im2ColTest.cpp]: Done.

File: paddle/function/GemmConvOp.cpp (hunk 4 above)
Comment (hedaoyuan) [reply paddle/function/GemmConvOp.cpp]: Done. Add release the memory.

**Pull request merged**